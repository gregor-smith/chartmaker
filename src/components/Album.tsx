import React, { FC } from 'react'
import { css, cx } from 'emotion'
import { Album as AlbumDetails, formatAlbumTitle } from '../state'


type Props = {
    details: AlbumDetails | null
    sizeRem: number
}


const baseStyle = css({
    backgroundColor: 'white',
    margin: '0.15rem'
})


export const Album: FC<Props> = ({ details, sizeRem }) => {
    const style = cx(
        baseStyle,
        css({
            width: `${sizeRem}rem`,
            height: `${sizeRem}rem`
        })
    )
    const image = details === null
        ? null
        : <img src={details.image.smallURL} alt={formatAlbumTitle(details)}/>
    return (
        <div className={style}>
            {image}
        </div>
    )
}
