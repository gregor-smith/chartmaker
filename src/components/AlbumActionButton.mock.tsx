const mock: typeof import('@/components/AlbumActionButton') = {
    AlbumActionButton: props =>
        <div {...props} className='mock-album-action-button'/>
}


export const AlbumActionButton = mock.AlbumActionButton
