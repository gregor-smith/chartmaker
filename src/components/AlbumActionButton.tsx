import { h, FunctionComponent } from 'preact'
import { css } from 'emotion'


type Props = {
    onClick: () => void
    title: string
}


const style = css({
    background: 'none',
    border: 'none'
})


export const AlbumActionButton: FunctionComponent<Props> = props =>
    <button {...props} class={style}/>
