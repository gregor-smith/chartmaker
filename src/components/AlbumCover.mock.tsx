import { cx } from 'emotion'


const mock: typeof import('@/components/AlbumCover') = {
    AlbumCover: ({
        album,
        size,
        overlayClass,
        children,
        highlighted,
        ...props
    }) => {
        const json = JSON.stringify(album)
        const className = cx('mock-album-cover', overlayClass)
        return (
            <div {...props} className={className}>
                {`Album: ${json}`}
                {`Size: ${size}`}
                {`Highlighted: ${highlighted}`}
                {children}
            </div>
        )
    }
}


export const AlbumCover = mock.AlbumCover
