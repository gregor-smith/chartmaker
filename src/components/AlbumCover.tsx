import React, { FC, DragEvent } from 'react'
import { css, cx } from 'emotion'

import { TEXT_COLOUR, ALBUM_PADDING_SIZE, highlightBackgroundStyle } from '@/style'
import { Album } from '@/types'


export type AlbumCoverProps = {
    album: Album
    size: string
    onDragStart: (event: DragEvent<HTMLDivElement>) => void
    onDragOver?: (event: DragEvent<HTMLDivElement>) => void
    onDragEnter?: (event: DragEvent<HTMLDivElement>) => void
    onDrop?: (event: DragEvent<HTMLDivElement>) => void
    onMouseEnter?: () => void
    overlayClass?: string
    highlighted?: boolean
}


const baseContainerStyle = css({
    position: 'relative',
    background: TEXT_COLOUR,
    margin: ALBUM_PADDING_SIZE
})

const imageStyle = css({
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    maxWidth: '100%',
    height: '100%',
    objectFit: 'contain'
})

// the overlay is given an additional non-emotion class so it can be used as a
// selector in the container's style
const overlaySelector = 'overlay'

const baseOverlayStyle = cx(
    overlaySelector,
    imageStyle,
    css({ zIndex: 10 })
)


export const AlbumCover: FC<AlbumCoverProps> = ({
    album,
    size,
    onDragStart,
    onDragEnter,
    onDragOver,
    onDrop,
    onMouseEnter,
    children,
    overlayClass,
    highlighted
}) => {
    const overlayStyle = cx(baseOverlayStyle, overlayClass)
    const containerStyle = cx(
        baseContainerStyle,
        css({
            width: size,
            height: size,
            [`:not(:hover) .${overlaySelector}`]: {
                display: 'none'
            }
        }),
        highlighted === false
            ? highlightBackgroundStyle
            : undefined
    )

    let image: JSX.Element | undefined
    if (!album.placeholder) {
        image = (
            // lazy loading causes problems with html2canvas so keep it default
            <img className={imageStyle}
                src={album.url}
                alt={album.name}
                crossOrigin='anonymous'/>
        )
    }

    return (
        <div className={containerStyle}
                draggable={!album.placeholder}
                onDragStart={onDragStart}
                onDragEnter={onDragEnter}
                onDragOver={onDragOver}
                onDrop={onDrop}
                onMouseEnter={onMouseEnter}
                title={album.placeholder ? undefined : album.name}>
            {image}
            <div className={overlayStyle}>
                {children}
            </div>
        </div>
    )
}
