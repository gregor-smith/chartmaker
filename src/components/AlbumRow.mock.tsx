const mock: typeof import('@/components/AlbumRow') = {
    AlbumRow: ({ children }) =>
        <div className='mock-album-row'>
            {children}
        </div>
}


export const AlbumRow = mock.AlbumRow
