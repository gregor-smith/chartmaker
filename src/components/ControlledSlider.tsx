import { h, FunctionComponent } from 'preact'

import { inputStyle } from '../style'


type Props = {
    id?: string
    min: number
    max: number
    step: number
    value: number
    onChange: (value: number) => void
    disabled?: boolean
}


export const ControlledSlider: FunctionComponent<Props> = ({
    onChange,
    ...props
}) => {
    function change(event: Event) {
        event.preventDefault()
        const value = Number((event.currentTarget as any).value)
        onChange(value)
    }

    return <input {...props} class={inputStyle} type='range' onInput={change}/>
}
