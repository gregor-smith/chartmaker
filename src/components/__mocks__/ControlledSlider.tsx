import React, { FC, ChangeEvent } from 'react'

import { ControlledSliderProps } from '@/components/ControlledSlider'


export const ControlledSlider: FC<ControlledSliderProps> = ({
    id,
    min,
    max,
    step,
    value,
    disabled,
    onChange,
    children
}) => {
    function change(event: ChangeEvent<HTMLDivElement>) {
        onChange(Number((event.target as any).value))
    }

    return (
        <div id={id} className='mock-controlled-slider' onChange={change}>
            {`Min: ${min}`}
            {`Max: ${max}`}
            {`Step: ${step}`}
            {`Value: ${value}`}
            {`Disabled: ${disabled}`}
            {children}
        </div>
    )
}
