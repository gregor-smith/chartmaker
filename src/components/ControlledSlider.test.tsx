import { render } from 'react-dom'
import { act } from 'react-dom/test-utils'

import { ControlledSlider } from '@/components/ControlledSlider'

import { RenderContainer, ignore, fireEvent } from '../utils'


jest.mock('@/components/Label')


const container = new RenderContainer()


test.each([ true, false, undefined ])('renders labelled slider', disabled => {
    render(
        <ControlledSlider id='test-slider'
                min={1}
                max={3}
                step={1}
                value={1}
                disabled={disabled}
                onChange={ignore}>
            Test children
        </ControlledSlider>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})


test('change event calls onChange prop', () => {
    const mock = jest.fn<void, [ number ]>()

    render(
        <ControlledSlider id='test-slider'
                min={1}
                max={3}
                step={1}
                value={1}
                disabled={false}
                onChange={mock}>
            Test children
        </ControlledSlider>,
        container.element
    )

    act(() =>
        fireEvent(
            'change',
            container.element?.querySelector('input'),
            { target: { value: '2' } }
        )
    )

    expect(mock).toHaveBeenCalledTimes(1)
    expect(mock).toHaveBeenCalledWith(2)
})
