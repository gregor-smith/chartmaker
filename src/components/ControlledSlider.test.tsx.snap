// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders labelled slider 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0 {
  margin: 0 0.5rem;
  background: black;
  color: white;
  border: 1px solid white;
  padding: 0.25rem;
  width: 100%;
}

<div
  id="test-render-container"
>
  <div
    class="emotion-1"
  >
    <div
      class="mock-label"
    >
      Target: test-slider
      Test children
    </div>
    <input
      class="emotion-0"
      disabled=""
      id="test-slider"
      max="3"
      min="1"
      step="1"
      type="range"
      value="1"
    />
    <div
      class="mock-label"
    >
      Target: test-slider
      1
    </div>
  </div>
</div>
`;

exports[`renders labelled slider 2`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0 {
  margin: 0 0.5rem;
  background: black;
  color: white;
  border: 1px solid white;
  padding: 0.25rem;
  width: 100%;
}

<div
  id="test-render-container"
>
  <div
    class="emotion-1"
  >
    <div
      class="mock-label"
    >
      Target: test-slider
      Test children
    </div>
    <input
      class="emotion-0"
      id="test-slider"
      max="3"
      min="1"
      step="1"
      type="range"
      value="1"
    />
    <div
      class="mock-label"
    >
      Target: test-slider
      1
    </div>
  </div>
</div>
`;

exports[`renders labelled slider 3`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0 {
  margin: 0 0.5rem;
  background: black;
  color: white;
  border: 1px solid white;
  padding: 0.25rem;
  width: 100%;
}

<div
  id="test-render-container"
>
  <div
    class="emotion-1"
  >
    <div
      class="mock-label"
    >
      Target: test-slider
      Test children
    </div>
    <input
      class="emotion-0"
      id="test-slider"
      max="3"
      min="1"
      step="1"
      type="range"
      value="1"
    />
    <div
      class="mock-label"
    >
      Target: test-slider
      1
    </div>
  </div>
</div>
`;
